RFC: Should we enable `-pedantic`?

## Description

@parker-research has found that we have a few minor errors in our code base which can be found with `-pedantic`.

- #645 
- #646 

We also have fixed a few others:

- #363
- https://github.com/libcsp/libcsp/commit/ed71a2a80f60b8366b849eb8353fe14e007d8710

Should we enable `-pedantic` by default?

## Steps to Reproduce

**CMake**
1. `cmake -B builddir -DCMAKE_C_FLAGS="-pedantic"`
2. `ninja -C builddir`

**Meson**
1. `meson setup builddir -Dc_args=-pedantic`
2. `ninja -C builddir`

## Expected behavior

No error even with `-pedantic`

## Actual Behavior:

```
ninja: Entering directory `builddir'
[7/41] Building C object CMakeFiles/csp.dir/src/csp_conn.c.o
/home/yashi/src/libcsp/src/csp_conn.c: In function ‘csp_conn_print_table_str’:
/home/yashi/src/libcsp/src/csp_conn.c:390:52: warning: format ‘%p’ expects argument of type ‘void *’, but argument 5 has type ‘csp_conn_t *’ {aka ‘struct csp_conn_s *’} [-Wformat=]
  390 |                 snprintf(buf, sizeof(buf), "[%02u %p] S:%u, %u -> %u, %u -> %u (%u)\n",
      |                                                   ~^
      |                                                    |
      |                                                    void *
  391 |                                  i, conn, conn->state, conn->idin.src, conn->idin.dst,
      |                                     ~~~~            
      |                                     |
      |                                     csp_conn_t * {aka struct csp_conn_s *}
[25/41] Building C object CMakeFiles/csp.dir/src/arch/posix/csp_queue.c.o
/home/yashi/src/libcsp/src/arch/posix/csp_queue.c: In function ‘csp_queue_empty’:
/home/yashi/src/libcsp/src/arch/posix/csp_queue.c:46:16: warning: ISO C forbids ‘return’ with expression, in function returning void [-Wpedantic]
   46 |         return pthread_queue_empty(handle);
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/yashi/src/libcsp/src/arch/posix/csp_queue.c:45:6: note: declared here
   45 | void csp_queue_empty(csp_queue_handle_t handle) {
      |      ^~~~~~~~~~~~~~~
[41/41] Linking C shared library libcsp.so
```

## Environment Details:

Provide information about your environment, including:

- Build OS: Debian Sid
- Target OS: Linux
- libcsp commit hash: `develop (v1.6-1128-g4c4fd14c)`
- CMake version: 3.30.2
- Ninja version: 1.12.1

## Logs or console output from the very beginning to the end

```
❯ rm -rf builddir && cmake -B builddir -DCMAKE_C_FLAGS="-pedantic" && ninja -C builddir
-- The C compiler identification is GNU 14.2.0
-- The CXX compiler identification is GNU 14.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/lib/ccache/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/lib/ccache/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- No build type selected, default to MinSizeRel
-- Looking for include file sys/socket.h
-- Looking for include file sys/socket.h - found
-- Looking for include file arpa/inet.h
-- Looking for include file arpa/inet.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Found PkgConfig: /usr/bin/pkg-config (found version "1.8.1")
-- Checking for one of the modules 'yaml-0.1'
-- Found /usr/lib/x86_64-linux-gnu/libyaml.so 0.2.5
-- Checking for one of the modules 'libzmq'
-- Found /usr/lib/x86_64-linux-gnu/libzmq.so 4.3.5
-- Checking for one of the modules 'libsocketcan'
-- Found /usr/lib/x86_64-linux-gnu/libsocketcan.so 0.0.12
-- Configuring done (0.5s)
-- Generating done (0.0s)
-- Build files have been written to: /home/yashi/src/libcsp/builddir
ninja: Entering directory `builddir'
[7/41] Building C object CMakeFiles/csp.dir/src/csp_conn.c.o
/home/yashi/src/libcsp/src/csp_conn.c: In function ‘csp_conn_print_table_str’:
/home/yashi/src/libcsp/src/csp_conn.c:390:52: warning: format ‘%p’ expects argument of type ‘void *’, but argument 5 has type ‘csp_conn_t *’ {aka ‘struct csp_conn_s *’} [-Wformat=]
  390 |                 snprintf(buf, sizeof(buf), "[%02u %p] S:%u, %u -> %u, %u -> %u (%u)\n",
      |                                                   ~^
      |                                                    |
      |                                                    void *
  391 |                                  i, conn, conn->state, conn->idin.src, conn->idin.dst,
      |                                     ~~~~            
      |                                     |
      |                                     csp_conn_t * {aka struct csp_conn_s *}
[25/41] Building C object CMakeFiles/csp.dir/src/arch/posix/csp_queue.c.o
/home/yashi/src/libcsp/src/arch/posix/csp_queue.c: In function ‘csp_queue_empty’:
/home/yashi/src/libcsp/src/arch/posix/csp_queue.c:46:16: warning: ISO C forbids ‘return’ with expression, in function returning void [-Wpedantic]
   46 |         return pthread_queue_empty(handle);
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/yashi/src/libcsp/src/arch/posix/csp_queue.c:45:6: note: declared here
   45 | void csp_queue_empty(csp_queue_handle_t handle) {
      |      ^~~~~~~~~~~~~~~
[41/41] Linking C shared library libcsp.so
```

## Checklist

- [x] I've searched in "Issues" for similar issues, both open and closed.
- [x] I've used GitHub Flavored Markdown (GFM) to format my text.
- [x] I've avoided using screenshots for console output or logs
- [x] I've included all relevant details and steps to reproduce the issue.
- [x] I've provided information about my environment.
- [x] I've attached necessary logs or code snippet for minimal reproducible example.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -8,7 +8,8 @@ endif()
 add_library(csp)
 set_target_properties(csp PROPERTIES C_STANDARD 11)
 set_target_properties(csp PROPERTIES C_EXTENSIONS ON)
-target_compile_options(csp PRIVATE -Wall -Wextra)
+target_compile_options(csp PRIVATE -Wall -Wextra -Wpedantic
+  $<$<C_COMPILER_ID:Clang>:-Wno-gnu-zero-variadic-macro-arguments>)
 
 set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS Debug Release MinSizeRel)
 if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)

--- a/examples/csp_arch.c
+++ b/examples/csp_arch.c
@@ -17,7 +17,7 @@ int main(int argc, char * argv[]) {
 
 
     // clock
-    csp_timestamp_t csp_clock = {};
+    csp_timestamp_t csp_clock = {0};
     csp_clock_get_time(&csp_clock);
     assert(csp_clock.tv_sec != 0);
     csp_print("csp_clock_get_time(..) -> sec:nsec = %"PRIu32":%"PRIu32"\n", csp_clock.tv_sec, csp_clock.tv_nsec);

--- a/examples/csp_client.c
+++ b/examples/csp_client.c
@@ -64,7 +64,7 @@ static struct option long_options[] = {
     {0, 0, 0, 0}
 };
 
-void print_help() {
+void print_help(void) {
 	csp_print("Usage: csp_client [options]\n");
 	if (CSP_HAVE_LIBSOCKETCAN) {
 		csp_print(" -c <can-device>  set CAN device\n");

--- a/examples/csp_server.c
+++ b/examples/csp_server.c
@@ -113,7 +113,7 @@ static struct option long_options[] = {
     {0, 0, 0, 0}
 };
 
-void print_help() {
+void print_help(void) {
     csp_print("Usage: csp_server [options]\n");
 	if (CSP_HAVE_LIBSOCKETCAN) {
 		csp_print(" -c <can-device>  set CAN device\n");

--- a/meson.build
+++ b/meson.build
@@ -1,7 +1,7 @@
 project('csp', 'c', version: '2.1', license: 'LGPL', meson_version : '>=0.53.2', default_options : [
 	'c_std=gnu11',
 	'optimization=s',
-	'warning_level=2',
+	'warning_level=3',
 	'werror=true'])
 
 cc = meson.get_compiler('c')
@@ -65,6 +65,9 @@ csp_c_args = ['-Wshadow',
 	      '-Wwrite-strings',
 	      '-Wpointer-arith',
 	      '-Wno-unused-parameter']
+if cc.get_id() == 'clang'
+	csp_c_args += '-Wno-gnu-zero-variadic-macro-arguments'
+endif
 
 subdir('src')
 subdir('include/csp')

--- a/src/drivers/eth/eth_linux.c
+++ b/src/drivers/eth/eth_linux.c
@@ -32,7 +32,7 @@ int csp_eth_tx_frame(void * driver_data, csp_eth_header_t *eth_frame) {
     const eth_context_t * ctx = (eth_context_t*)driver_data;
 
     /* Destination socket address */
-    struct sockaddr_ll socket_address = {};
+    struct sockaddr_ll socket_address = {0};
     socket_address.sll_ifindex = ctx->if_idx.ifr_ifindex;
     socket_address.sll_halen = CSP_ETH_ALEN;
     memcpy(socket_address.sll_addr, eth_frame->ether_dhost, CSP_ETH_ALEN);

--- a/src/interfaces/csp_if_eth.c
+++ b/src/interfaces/csp_if_eth.c
@@ -77,7 +77,7 @@ arp_list_entry_t * arp_alloc(void) {
 
 }
 
-void arp_print()
+void arp_print(void)
 {
     csp_print("ARP  CSP  MAC\n");
     for (arp_list_entry_t * arp = arp_list; arp; arp = arp->next) {

--- a/wscript
+++ b/wscript
@@ -77,8 +77,10 @@ def configure(ctx):
     # Setup CFLAGS
     if (len(ctx.stack_path) <= 1) and (len(ctx.env.CFLAGS) == 0):
         ctx.env.prepend_value('CFLAGS', ["-std=gnu11", "-g", "-Os", "-Wall", "-Wextra", "-Wshadow", "-Wcast-align",
-                                         "-Wpointer-arith",
+                                         "-Wpointer-arith", "-Wpedantic",
                                          "-Wwrite-strings", "-Wno-unused-parameter", "-Werror"])
+        if ctx.env.CC_NAME == 'clang':
+            ctx.env.append_value('CFLAGS', ["-Wno-gnu-zero-variadic-macro-arguments"])
 
     # Setup default include path and any extra defined
     ctx.env.append_unique('INCLUDES_CSP', ['include', 'src'] + ctx.options.includes.split(','))

